common/src/main/java/io/netty/util/DefaultAttributeMap.java

```diff
@@ -17,186 +17,240 @@
+     * save polymorphic calls (on comparator side) and unnecessary class checks.
+     */
+    private static int searchAttributeByKey(DefaultAttribute[] sortedAttributes, int fromIndex, int toIndex,
+                                            AttributeKey<?> key) {
+        int low = fromIndex;
+        int high = toIndex - 1;
 
-    // Initialize lazily to reduce memory consumption; updated by AtomicReferenceFieldUpdater above.
-    @SuppressWarnings("UnusedDeclaration")
-    private volatile AtomicReferenceArray<DefaultAttribute<?>> attributes;
-
-    @SuppressWarnings("unchecked")
-    @Override
-    public <T> Attribute<T> attr(AttributeKey<T> key) {
-        ObjectUtil.checkNotNull(key, "key");
-        AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
-        if (attributes == null) {
-            // Not using ConcurrentHashMap due to high memory consumption.
-            attributes = new AtomicReferenceArray<DefaultAttribute<?>>(BUCKET_SIZE);
-
-            if (!updater.compareAndSet(this, null, attributes)) {
-                attributes = this.attributes;
+        while (low <= high) {
+            int mid = low + high >>> 1;
+            DefaultAttribute midVal = sortedAttributes[mid];
+            AttributeKey midValKey = midVal.key;
+            if (midValKey == key) {
+                assert midValKey.id() == key.id();
+                return mid;
+            }
+            int midValKeyId = midValKey.id();
+            int keyId = key.id();
+            assert midValKeyId != keyId;
+            boolean searchRight = midValKeyId < keyId;
+            if (searchRight) {
+                low = mid + 1;
+            } else {
+                high = mid - 1;
             }
         }
 
-        int i = index(key);
-        DefaultAttribute<?> head = attributes.get(i);
-        if (head == null) {
-            // No head exists yet which means we may be able to add the attribute without synchronization and just
-            // use compare and set. At worst we need to fallback to synchronization and waste two allocations.
-            head = new DefaultAttribute();
-            DefaultAttribute<T> attr = new DefaultAttribute<T>(head, key);
-            head.next = attr;
-            attr.prev = head;
-            if (attributes.compareAndSet(i, null, head)) {
-                // we were able to add it so return the attr right away
-                return attr;
-            } else {
-                head = attributes.get(i);
+        return -(low + 1);
+    }
+
+    private static void orderedCopyOnInsert(DefaultAttribute[] sortedSrc, int srcLength, DefaultAttribute[] copy,
+                                            DefaultAttribute toInsert) {
+        // let's walk backward, because as a rule of thumb, toInsert.key.id() tends to be higher for new keys
+        final int id = toInsert.key.id();
+        int i;
+        for (i = srcLength - 1; i >= 0; i--) {
```

Comment:
might be a nit, but did you consider binary searching to find the insertion point, and then use System.arraycopy to move the daat around for the slot?  I am curious if this more easily intrinsifies this for loop copy.

Reply:
As a general rule of thumb, yes, you're right, here I have taken advantage of the domain logic where latest addition of attributes tends to have higher IDs (due to the incremental logic on constants).